x-pack/plugins/cases/server/connectors/cases/cases_connector_executor.test.ts

```diff
@@ -0,0 +1,2086 @@
+            statusCode: 409,
+            error: 'Conflict',
+          },
+        ])
+        .mockResolvedValueOnce([{ ...oracleRecords[0], counter: 2 }]);
+
+      await expect(() =>
+        connectorExecutor.execute({
+          alerts,
+          groupingBy,
+          owner,
+          rule,
+          timeWindow,
+          reopenClosedCases,
+        })
+      ).rejects.toThrowErrorMatchingInlineSnapshot(
+        `"Conflict: updating records: mockBulkUpdateRecord error"`
+      );
+
+      resetCounters();
+
+      mockGetCaseId
+        .mockReturnValueOnce('mock-id-4')
+        .mockReturnValueOnce('mock-id-1')
+        .mockReturnValueOnce('mock-id-2');
+
+      casesClientMock.cases.bulkGet.mockResolvedValue({
+        cases: [cases[0], cases[1]],
+        errors: [
+          {
+            error: 'Not found',
+            message: 'Not found',
+            status: 404,
+            caseId: 'mock-id-4',
+          },
+        ],
+      });
+
+      casesClientMock.cases.bulkCreate.mockResolvedValue({
+        cases: [{ ...cases[0], id: 'mock-id-4' }],
+      });
+
+      // retry
+      await connectorExecutor.execute({
+        alerts,
+        groupingBy,
+        owner,
+        rule,
+        timeWindow,
+        reopenClosedCases,
+      });
+
+      expectCasesToHaveTheCorrectAlertsAttachedWithGroupingAndIncreasedCounter(casesClientMock);
+    });
+
+    it('attaches the alerts correctly when increasing the counter (time window) and another node has already increased it', async () => {
+      dateMathMock.parse.mockImplementation(() => moment('2023-10-11T10:23:42.769Z'));
+
+      mockBulkGetRecords
+        // counter is 1
+        .mockResolvedValueOnce([oracleRecords[0], oracleRecords[1]])
+        .mockResolvedValueOnce([
+          {
+            ...createdOracleRecord,
+            // another node increased the counter
+            counter: 2,
+            id: groupedAlertsWithOracleKey[0].oracleKey,
+            grouping: groupedAlertsWithOracleKey[0].grouping,
+            version: 'so-version-3',
+            createdAt: '2023-11-13T10:23:42.769Z',
+            updatedAt: '2023-11-13T10:23:42.769Z',
+          },
+          oracleRecords[1],
+        ]);
+
+      // conflict error. Another node had updated the record.
+      mockBulkUpdateRecord.mockResolvedValueOnce([
+        {
+          id: groupedAlertsWithOracleKey[0].oracleKey,
+          type: CASE_ORACLE_SAVED_OBJECT,
+          message: 'updating records: mockBulkUpdateRecord error',
+          statusCode: 409,
+          error: 'Conflict',
+        },
+      ]);
+
+      await expect(() =>
+        connectorExecutor.execute({
+          alerts,
+          groupingBy,
+          owner,
+          rule,
+          timeWindow,
+          reopenClosedCases,
+        })
+      ).rejects.toThrowErrorMatchingInlineSnapshot(
+        `"Conflict: updating records: mockBulkUpdateRecord error"`
+      );
```

Comment:
I get that since we are testing retries we want to actually call the same thing(`connectorExecutor.execute`) twice.

How relevant is that though?

We always have the same chunks in these tests. We mock some error for `mockBulk*Record` and expect some snapshot for `connectorExecutor.execute`. Are we really ensuring something or is this just overhead?

I don't know, food for thought. Maybe some integration tests would be more useful.

Reply:
The execution of the connector is designed to be as idempotent as it can be. Retrying should not affect the proper execution of the connector, alerts should always attached to the correct cases. In the RFC you can see how the connector is simulated as a state machine where each error and retry leads to the correct state each time, at least in theory. The tests try to simulate multiple nodes making changes to cases at the same time (race conditions) which unfortunately cannot be tested with integration tests. When the system actions are ready we will write a lot of integration tests but for the logic of one node executing the connector.

To the tests now, if you see the mock functions that we are interested in have a chain of `mockResolvedValueOnce(...). mockResolvedValueOnce(...)`. The first `mockResolvedValueOnce` will be called on the first execution and the second one on the second. This way we try to simulate different responses based on different actions on different nodes. For example, if the first `mockResolvedValueOnce` fails (conflict), then we retry, and on the second time return different results we test that on the first try another node did some change (increased the counter for example, or created the case for us) and on the second try we get the new state. The new state should not affect the execution of the connector and attach the alerts to the correct case. The snapshot check is to ensure that the correct function threw the error and not another function. After checking the snapshot we check in which case the alerts got attached.